models.py

```diff
@@ -323,32 +328,58 @@ def get_chronological(self, limit=None, update_cache=False):
     feature_list = memcache.get(KEY)
 
     if feature_list is None or update_cache:
-      q = Feature.all()
-      q.order('-shipped_milestone')
-      q.order('name')
-      features = q.fetch(None)
-
-      features = [f for f in features if (IN_DEVELOPMENT < f.impl_status_chrome < NO_LONGER_PURSUING)]
 
-      # Append no active, in dev, proposed features.
+      # Features with no active, in dev, proposed features.
       q = Feature.all()
       q.order('impl_status_chrome')
       q.order('name')
       q.filter('impl_status_chrome <=', IN_DEVELOPMENT)
       pre_release = q.fetch(None)
-      pre_release.extend(features)
 
-      # Append no longer pursuing features.
+      # Shipping features. Exclude features that do not have a desktop
+      # shipping milestone.
+      q = Feature.all()
+      q.order('-shipped_milestone')
+      q.order('name')
+      q.filter('shipped_milestone !=', None)
+      shipping_features = q.fetch(None)
+
+      # Features with an android shipping milestone but no desktop milestone.
+      q = Feature.all()
+      q.order('-shipped_android_milestone')
+      q.order('name')
+      q.filter('shipped_milestone =', None)
+      android_only_shipping_features = q.fetch(None)
+
+      # No longer pursuing features.
       q = Feature.all()
       q.order('impl_status_chrome')
       q.order('name')
       q.filter('impl_status_chrome =', NO_LONGER_PURSUING)
-      no_longer_pursuing = q.fetch(None)
-      pre_release.extend(no_longer_pursuing)
+      no_longer_pursuing_features = q.fetch(None)
 
-      feature_list = [f.format_for_template() for f in pre_release]
+      shipping_features.extend(android_only_shipping_features)
 
-      memcache.set(KEY, feature_list)
+      shipping_features = [f for f in shipping_features if (IN_DEVELOPMENT < f.impl_status_chrome < NO_LONGER_PURSUING)]
+
+      def getSortingMilestone(feature):
+        feature._sort_by_milestone = (feature.shipped_milestone or
+                                      feature.shipped_android_milestone)
+        return feature
+
+      # Sort the feature list on either Android shipping milestone or desktop
+      # shipping milestone, depending on which is specified. If a desktop
+      # milestone is defined, that will take default.
+      shipping_features = map(getSortingMilestone, shipping_features)
+
+      # TODO: sort on name after sorting on milestone.
```

Comment:
Any reason not to add the code to sort by name into this PR? You might be able to just put together a synthetic sort key that combines the version, a separator string, and then the name, and keep most of the code the same:

```python
shipping_features.sort(key=lambda f: "%s$$$%s" % (f._sort_by_milestone, f.name), reverse=True)
```